packages/expo-updates/android/src/main/java/expo/modules/updates/loader/FileDownloader.kt

```diff
@@ -82,6 +84,157 @@ open class FileDownloader(context: Context) {
     )
   }
 
+  internal fun parseManifestResponse(response: Response, configuration: UpdatesConfiguration, callback: ManifestDownloadCallback) {
+    val contentType = response.header("content-type") ?: ""
+    val isMultipart = contentType.startsWith("multipart/", ignoreCase = true)
+    if (isMultipart) {
+      val boundaryParameter = ParameterParser().parse(contentType, ';')["boundary"]
+      if (boundaryParameter == null) {
+        callback.onFailure(
+          "Missing boundary in multipart manifest content-type",
+          IOException("Missing boundary in multipart manifest content-type")
+        )
+        return
+      }
+
+      parseMultipartManifestResponse(response, boundaryParameter, configuration, callback)
+    } else {
+      parseManifest(response.body()!!.string(), response.headers(), null, configuration, callback)
+    }
+  }
+
+  private fun parseHeaders(text: String): Headers {
+    val headers = mutableMapOf<String, String>()
+    val lines = text.split(CRLF)
+    for (line in lines) {
+      val indexOfSeparator = line.indexOf(":")
+      if (indexOfSeparator == -1) {
+        continue
+      }
+      val key = line.substring(0, indexOfSeparator).trim()
+      val value = line.substring(indexOfSeparator + 1).trim()
+      headers[key] = value
+    }
+    return Headers.of(headers)
+  }
+
+  private fun parseMultipartManifestResponse(response: Response, boundary: String, configuration: UpdatesConfiguration, callback: ManifestDownloadCallback) {
+    var manifestBodyAndHeaders: Pair<String, Headers>? = null
+    var extensionsBody: String? = null
+
+    val multipartStream = MultipartStream(response.body()!!.byteStream(), boundary.toByteArray())
+
+    try {
+      var nextPart = multipartStream.skipPreamble()
+      while (nextPart) {
```

Comment:
I know this is how https://commons.apache.org/proper/commons-fileupload/apidocs/org/apache/commons/fileupload/MultipartStream.html is written but I don't get how the first boundary delimiter is consumed. skipPreamble() finds the first "encapsulation", which consists of a boundary + headers + body.

Reply:
Reading the code it looks like it reads up to the first delimiter and then consumes the delimiter. That being said the way these classes are written is hard to interpret. I'll trust our test to verify that the preamble is correctly skipped whether present or not.

Reply:
The "finally" block looked like it unconsumed the delimiter but I didn't stare at the arraycopy method long enough